mmedit/datasets/pipelines/blur_kernels.py

```diff
@@ -0,0 +1,519 @@
+# This code is referenced from BasicSR with modifications.
+# Reference: https://github.com/xinntao/BasicSR/blob/master/basicsr/data/degradations.py  # noqa
+import random
+
+import numpy as np
+from scipy import special
+
+np.seterr(divide='ignore', invalid='ignore')
+
+
+def sigma_matrix2(sig_x, sig_y, theta):
+    """Calculate the rotated sigma matrix (two dimensional matrix).
+
+    Args:
+        sig_x (float): Standard deviation along the horizontal direction.
+        sig_y (float): Standard deviation along the vertical direction.
+        theta (float): Rotation in radian.
+
+    Returns:
+        ndarray: Rotated sigma matrix.
+    """
+
+    diag = np.array([[sig_x**2, 0], [0, sig_y**2]])
+    rot = np.array([[np.cos(theta), -np.sin(theta)],
+                    [np.sin(theta), np.cos(theta)]])
+
+    return np.dot(rot, np.dot(diag, rot.T))
+
+
+def mesh_grid(kernel_size):
+    """Generate the mesh grid, centering at zero.
+
+    Args:
+        kernel_size (int): The size of the kernel.
+
+    Returns:
+        x_grid (ndarray): x-coordinates with shape (kernel_size, kernel_size).
+        y_grid (ndarray): y-coordiantes with shape (kernel_size, kernel_size).
+        xy_grid (ndarray): stacked coordinates with shape
+            (kernel_size, kernel_size, 2).
+    """
+
+    range_ = np.arange(-kernel_size // 2 + 1., kernel_size // 2 + 1.)
+    x_grid, y_grid = np.meshgrid(range_, range_)
+    xy_grid = np.hstack((x_grid.reshape((kernel_size * kernel_size, 1)),
+                         y_grid.reshape(kernel_size * kernel_size,
+                                        1))).reshape(kernel_size, kernel_size,
+                                                     2)
+
+    return xy_grid, x_grid, y_grid
+
+
+def calculate_gaussian_pdf(sigma_matrix, grid):
+    """Calculate PDF of the bivariate Gaussian distribution.
+
+    Args:
+        sigma_matrix (ndarray): The variance matrix with shape (2, 2).
+        grid (ndarray): Coordinates generated by :func:`mesh_grid`,
+            with shape (K, K, 2), where K is the kernel size.
+
+    Returns:
+        kernel (ndarrray): Un-normalized kernel.
+    """
+
+    inverse_sigma = np.linalg.inv(sigma_matrix)
+    kernel = np.exp(-0.5 * np.sum(np.dot(grid, inverse_sigma) * grid, 2))
```

Comment:
The behavior of np.dot is not very consistent see https://numpy.org/doc/stable/reference/generated/numpy.dot.html. So if you meant mat mult, then use matmul.

A guideline is to avoid using np.dot if possible